src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+        }
+    } else if (serverEntry.isDirectory != (dbEntry._type == ItemTypeDirectory)) {
+        // If the type of the entity changed, it's like NEW, but
+        // needs to delete the other entity first.
+        item->_instruction = CSYNC_INSTRUCTION_TYPE_CHANGE;
+        item->_direction = SyncFileItem::Down;
+        item->_modtime = serverEntry.modtime;
+        item->_size = serverEntry.size;
+    } else if (dbEntry._type == ItemTypeVirtualFileDownload) {
+        item->_direction = SyncFileItem::Down;
+        item->_instruction = CSYNC_INSTRUCTION_NEW;
+        item->_file = _currentFolder._target + QLatin1Char('/') + serverEntry.name;
+        item->_type = ItemTypeVirtualFileDownload;
+    } else if (dbEntry._etag != serverEntry.etag) {
+        item->_direction = SyncFileItem::Down;
+        item->_modtime = serverEntry.modtime;
+        item->_size = serverEntry.size;
+        if (serverEntry.isDirectory && dbEntry._type == ItemTypeDirectory) {
+            item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+        } else if (!localEntry.isValid() && _queryLocal != ParentNotChanged) {
+            // Deleted locally, changed on server
+            item->_instruction = CSYNC_INSTRUCTION_NEW;
+        } else {
+            item->_instruction = CSYNC_INSTRUCTION_SYNC;
+        }
+    } else if (dbEntry._remotePerm != serverEntry.remotePerm || dbEntry._fileId != serverEntry.fileId) {
+        item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+        item->_direction = SyncFileItem::Down;
+    } else {
+        processFileAnalyzeLocalInfo(item, path, localEntry, serverEntry, dbEntry, ParentNotChanged);
+        return;
+    }
+
+    processFileAnalyzeLocalInfo(item, path, localEntry, serverEntry, dbEntry, _queryServer);
+}
+
+void ProcessDirectoryJob::processFileAnalyzeLocalInfo(
+    const SyncFileItemPtr &item, PathTuple path, const LocalInfo &localEntry,
+    const RemoteInfo &serverEntry, const SyncJournalFileRecord &dbEntry, QueryMode recurseQueryServer)
+{
+    bool noServerEntry = (_queryServer != ParentNotChanged && !serverEntry.isValid())
+        || (_queryServer == ParentNotChanged && !dbEntry.isValid());
+
+    if (noServerEntry)
+        recurseQueryServer = ParentDontExist;
+
+    bool serverModified = item->_instruction == CSYNC_INSTRUCTION_NEW || item->_instruction == CSYNC_INSTRUCTION_SYNC
+        || item->_instruction == CSYNC_INSTRUCTION_RENAME || item->_instruction == CSYNC_INSTRUCTION_TYPE_CHANGE;
+
+    if ((dbEntry.isValid() && dbEntry._type == ItemTypeVirtualFile) || (localEntry.isValid() && localEntry.isVirtualFile && item->_type != ItemTypeVirtualFileDownload)) {
+        // Do not download virtual files
+        if (serverModified || dbEntry._type != ItemTypeVirtualFile)
+            item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+        serverModified = false;
+        item->_type = ItemTypeVirtualFile;
+    }
+    if (!_dirItem || _dirItem->_direction == SyncFileItem::Up) {
```

Comment:
I did not know either :-). So I removed the condition to see what happens, and it turned out it is because of CSYNC_INSTRUCTION_TYPE_CHANGE.  In this case, we do not want to mark it as a conflict if the new directory has new files (which is to be expected)

I now removed the condition and checked it in another way.  Only when removing a directory I check if _childModified

Notice that while doing that I realized that we will delete the directory and proceed with the type change if the directory being removed contains ignored item...
Maybe this can be improved...

Reply:
I'm sceptical as there's lots of code like this - where _childModified gets changed after some Up/Down check. I've just never dug into what's going on exactly. Guarding against deleting stuff with ignores or other changes on the opposide side makes a lot of sense.